Create only 1 workspace when signing out from an unvalidated account and signing into a new one.

neil-marcellini opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->
Requesting a review from @TomatoToaster since he helped me think through this one.
### Details
<!-- Explanation of the change or anything fishy that is going on -->
When userB lands on the [LogInWithShortLivedTokenPage](https://github.com/Expensify/App/blob/ec4df0da935353980a079ccbd0b9a302b9e1c70e/src/pages/LogInWithShortLivedTokenPage.js) after they have already been through this flow with userA, they will now be signed out and signed in as userB only once. One workspace will be created for them. 

Previously the user was signed out and signed back in multiple times because the network requests from the old userA were not cleared when calling [Session.signOut()](https://github.com/Expensify/App/blob/ec4df0da935353980a079ccbd0b9a302b9e1c70e/src/libs/actions/Session/index.js#L74). Logging in and out multiple times caused multiple workspaces to be created.

Please see the original issue for more context.
### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7617

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Open an incognito window (so that there are no cookies or local data) and go to expensify.com.dev.
2. Sign in to expensify.com.dev as a brand new userA with an `@gmail.com` email address.
3. Click to open the second inbox task that says "Would you like to get started with our free plan?".
4. Click get started.
5. Verify and note that userA is now logged into NewDot.
6. Navigate back to expensify.com.dev and sign out as userA.
7. Sign in to expensify.com.dev as a brand new userB with an `@gmail.com` email address.
8. Click to open the second inbox task that says "Would you like to get started with our free plan?".
9. Click get started.
10. Verify and note that User B is now logged into NewDot.
11. Go back to expensify.com.dev and go to Settings>Policies>Group.
12. Refresh the page.
13. Verify that there is 1 workspace policy.

See the original issue for a video showing the testing.

- [x] Verify that no new errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I made sure to pull `main` before submitting my PR for review
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I clearly indicated the environment tests should be run in (Staging vs Production)
- [x] I wrote testing steps that cover success & fail scenarios (if applicable)
- [x] I ran the tests & they passed on **all platforms**
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified there are no console errors related to changes in this PR
- [x] I followed proper code patterns (see [Reviewing the code](../PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I added comments when the code was not self explanatory
    - [x] I put all copy / text shown in the product in all `src/languages/*` files (if applicable)
    - [x] I followed proper naming convention for platform-specific files (if applicable)
    - [x] I followed style guidelines (in [`Styling.md`](../STYLING.md)) for all style edits I made
- [x] I followed the guidelines as stated in the [Review Guidelines](../PR_REVIEW_GUIDELINES.md)

#### PR Reviewer Checklist
- [ ] I verified the Author pulled `main` before submitting the PR
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the testing environment is mentioned in the test steps
- [ ] I verified testing steps cover success & fail scenarios (if applicable)
- [ ] I verified tests pass on **all platforms** & I tested again on all platforms
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified there are no console errors related to changes in this PR
- [ ] I verified proper code patterns were followed (see [Reviewing the code](../PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified comments were added when the code was not self explanatory
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [ ] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [ ] I verified [style guidelines](../STYLING.md) were followed
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](../PR_REVIEW_GUIDELINES.md)

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as tests but on staging.
- [ ] Verify that no errors appear in the JS console

### Tested On

- [ ] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/26260477/155796320-f1090cbc-4357-46bf-8818-cbc0b7e0ff7c.mov


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
![mWeb](https://user-images.githubusercontent.com/26260477/155799089-568a304c-6a6d-4b27-9006-d00cfc85f456.png)

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
<img width="1312" alt="image" src="https://user-images.githubusercontent.com/26260477/155815301-48dc3fcb-b071-4f39-abb9-3953a620e99b.png">

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
![ios](https://user-images.githubusercontent.com/26260477/155815320-eaa8eebf-797e-4c91-b348-e698ef59ef28.png)

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
<img width="551" alt="image" src="https://user-images.githubusercontent.com/26260477/155815846-53e5a1b8-3216-46a0-840f-6783b7b7296a.png">

neil-marcellini committed:
Sign out and sign in only once

marcaaron commented:
Please no! I never want to look at this code again! :trollface:

neil-marcellini committed:
Add account to defaultProps

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -57,9 +67,14 @@ class LogInWithShortLivedTokenPage extends Component {
         const email = lodashGet(this.props.route.params, 'email', '');
         const shortLivedToken = lodashGet(this.props.route.params, 'shortLivedToken', '');
 
+        // If the account is loading don't send more requests
```
This comment is ambiguous and made inspired a few questions...

- which requests we don't want to send?
- how does the early return "stop" those requests from sending?
- why is the account in a `loading` state?

TomatoToaster replied:
Yes I think we should specify here that the account is already trying to authenticate and that we don't want to reattempt authenticating.

neil-marcellini committed:
Sign out and sign in separately

neil-marcellini commented:
I updated this so that I don't need to check if the account is loading. If the user is signed in, then I sign them out. When this page loads as a public screen because the user is not signed in, then it will go ahead and sign them in. I also updated the code in componentDidUpdate, but it is not called with this flow.

neil-marcellini committed:
Merge branch 'main' of github.com:Expensify/App into neil-welcome-free-workspace

TomatoToaster commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -57,19 +57,16 @@ class LogInWithShortLivedTokenPage extends Component {
-        // If the user is revisiting the component authenticated with the right account, we don't need to do anything, the componentWillUpdate when betas are loaded and redirect
-        if (this.props.session.authToken && email === this.props.session.email) {
-            return;
+        // User is trying to transition with a different account than the one they are currently signed in as so we will sign out and then sign in with the new authToken
```
NAB: Could be helpful to add that they will be signed in because they will be navigated back to this component as a public screen. It could be confusing for someone reading this in the future to figure out how this is happening.

neil-marcellini replied:
Good idea, I updated the comment.

neil-marcellini committed:
Explain transitioning to a new account

neil-marcellini commented:
Updated with a more detailed comment.

neil-marcellini committed:
Update react-native-onyx version

Prevent setting state on unmounted withOnyx components

neil-marcellini commented:
Updated to use react-native-onyx after https://github.com/Expensify/react-native-onyx/pull/120

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -57,19 +57,20 @@ class LogInWithShortLivedTokenPage extends Component {
-        // User is trying to transition with a different account than the one they are currently signed in as so we will sign out and then sign in with the new authToken
-        if (email !== this.props.session.email) {
-            Session.signOut();
+        const signedIn = this.props.session && this.props.session.authToken;
+        if (!signedIn) {
```
```suggestion
        const isUserSignedIn = this.props.session && this.props.session.authToken;
        if (!isUserSignedIn) {
```
to keep variable naming consistent with what is written here https://github.com/Expensify/App/blob/main/STYLE.md#boolean-variables-and-props

neil-marcellini replied:
Ah thanks for pointing that out!

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -82,15 +83,24 @@ class LogInWithShortLivedTokenPage extends Component {
-        // User is signing in to a different account sign them out and sign in again
-        if (email !== this.props.session.email) {
+        const signedIn = this.props.session && this.props.session.authToken;
+        if (!signedIn) {
```
Same comment about variable naming

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -82,15 +83,24 @@ class LogInWithShortLivedTokenPage extends Component {
+        if (email !== this.props.session.email) {
+            Session.signOutAndRedirectToSignIn();
+            return;
+        }
```
Small idea on how to improve this. I noticed that the comment here is pretty similar to the one in `componentDidMount()`. What if we did something like this...

```js
if (this.signOutIfNeeded(email)) {
    return;
}
```

And then wrote a method that does this

```js
// add the long comment here in the JSDoc...
signOutIfNeeded(email) {
    if (email === this.props.session.email) {
        return false;
    }

    Session.signOutAndRedirectToSignIn();
    return true;
}
```

neil-marcellini replied:
That's a great idea, I'll do that.

neil-marcellini replied:
I also extracted the sign in logic to a function because it was also repeated twice.

TomatoToaster commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -57,19 +57,20 @@ class LogInWithShortLivedTokenPage extends Component {
             return;
         }
 
-        // If the user is revisiting the component authenticated with the right account, we don't need to do anything, the componentWillUpdate when betas are loaded and redirect
```
Actually I just realized this, could you also get rid of the `betas` in this file? This is the last comment explaining why they're here and it's not really relevant to keep the betas in withOnyx, componentDidMount, or propTypes/default.

neil-marcellini replied:
Ok that's a good catch. It doesn't really look like they are used. I'll remove betas and test it out.

marcaaron replied:
>the componentWillUpdate when betas are loaded and redirect

I guess that was important because the component stopped updating 😄

neil-marcellini committed:
Handle sign out and in with functions

neil-marcellini committed:
Remove check for betas

neil-marcellini commented:
Updated!

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -38,59 +38,40 @@ const propTypes = {
-
-        // exitTo is URI encoded because it could contain a variable number of slashes (i.e. "workspace/new" vs "workspace/<ID>/card")
-        const exitTo = decodeURIComponent(lodashGet(this.props.route.params, 'exitTo', ''));
+        if (this.signInIfNeeded()) {
```
I think this cleanup would maybe would be better in a separate follow up PR. 
Was pretty close to merging this code, but honestly not too sure about the latest changes...

marcaaron replied:
1. It makes this change harder to track and understand in isolation. 
2. I find it strange that `componentDidMount()` was signing out first then signing in. And `componentDidUpdate()` never needed a check to sign in and we are now adding one... so I question the necessity of that.

Basically, I have to think about these changes more (I don't really like thinking more than I have to) and would rather think about whether this component needs to be improved further or not (I'm not sure it does).

neil-marcellini replied:
> 1. It makes this change harder to track and understand in isolation.

I agree. The same is true for adding a signOutIfNeeded function. I could do both of these in a follow up PR.

> 2. I find it strange that `componentDidMount()` was signing out first then signing in (now looks to be doing the opposite). And `componentDidUpdate()` never needed a check to "sign in" and we are now adding one... so I question the necessity of that.

componentDidMount() has been signing in first and then signing out, but really it only does on or the other. [This](https://github.com/Expensify/App/blob/4fbbdd83e0e93eca1b749210b8d8f036ba596b25/src/pages/LogInWithShortLivedTokenPage.js#L60-L64) is what I had before the refactor.

componentDidUpdate() always had a check to sign in, both [before all my changes](https://github.com/Expensify/App/blob/4cbd7ff86cfed6214a01231208fa2555269b403b/src/pages/LogInWithShortLivedTokenPage.js#L101-L107) and [before my refactor](https://github.com/Expensify/App/blob/4fbbdd83e0e93eca1b749210b8d8f036ba596b25/src/pages/LogInWithShortLivedTokenPage.js#L86). That being said I'm not totally sure if it is necessary to check if the user is signed in here.

Please let me know what you think. If this still seems like too many changes then I'll go back to what I had in commit 39bd7d5f977d11074552a2a3a3194a14f225531f, and I'll keep the commit with the removal of the betas.

neil-marcellini committed:
Revert "Handle sign out and in with functions"

This reverts commit 8268ef406f1747dd408a0f2c72d388186c7dff6f.

neil-marcellini committed:
Extract only the duplicated sign out logic

neil-marcellini commented:
Updated to undo my refactor and only extract the sign out logic to a function.

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -82,15 +72,19 @@ class LogInWithShortLivedTokenPage extends Component {
-        // User is signing in to a different account sign them out and sign in again
-        if (email !== this.props.session.email) {
+        const signedIn = this.props.session && this.props.session.authToken;
+        if (!signedIn) {
```
Couple things... 

1. Not sure I understand why we need to handle the case of a user who is not signed in here (looks like that is covered in `componentDidMount()`). Is it possible for the user to be "not signed in" at this point?
2. Looks like we are back to using `signedIn` instead of `isUserSignedIn` or something equivalent.

neil-marcellini replied:
Great points, I should have caught those things. The are times that componentDidUpdate runs when the user is signed out, and we do not want to create a duplicate call to create a login. So how about this:

```
componentDidUpdate() {
     const isUserSignedIn = this.props.session && this.props.session.authToken;
      if (!isUserSignedIn) {
          return;
      }

      const email = lodashGet(this.props.route.params, 'email', '');

      // exitTo is URI encoded because it could contain a variable number of slashes (i.e. "workspace/new" vs "workspace/<ID>/card")
      const exitTo = decodeURIComponent(lodashGet(this.props.route.params, 'exitTo', ''));
     ...
}
```

marcaaron replied:
Yep, looks like we're already doing this to prevent it

https://github.com/Expensify/App/blob/bcc7630a3fa40ca2a94ec674ccaf75862a697d8d/src/pages/LogInWithShortLivedTokenPage.js#L76-L78

So changing it to this makes sense

```js
if (!this.props.session.authToken) {
    return;
}
```

neil-marcellini committed:
Remove sign in check from componentDidUpdate

neil-marcellini commented:
Updated again.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1965977435) to staging by @Justicea83 in version: 1.1.42-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1989053438) to production by @chiragsalian in version: 1.1.42-6 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

marcaaron commented on src/pages/LogInWithShortLivedTokenPage.js:
```diff
@@ -118,7 +121,4 @@ export default withOnyx({
     },
-    betas: {
-        key: ONYXKEYS.BETAS,
-    },
```
I think removing these betas maybe broke the transition when we are signed in because `componentDidUpdate()` never runs now. It's not really obvious, but this comment here that we deleted seems like a clue: 

https://github.com/Expensify/App/pull/7902/files#diff-94f637f52b69835bc8de01012bdb6ca75a5d07ba887bc0c9e7eb00e0a17186cbL67

This flow is so painful... we should just re-write it
